db/repair_bucket.go

```diff
@@ -0,0 +1,228 @@
+package db
+
+import (
+	"encoding/json"
+
+	"io/ioutil"
+
+	"fmt"
+
+	"github.com/couchbase/go-couchbase"
+	"github.com/couchbase/sync_gateway/base"
+)
+
+// Enum for the different repair jobs (eg, repairing rev tree cycles)
+type RepairJobType string
+
+const (
+	RepairRevTreeCycles = RepairJobType("RepairRevTreeCycles")
+)
+
+// Params suitable for external (eg, HTTP) invocations to describe a RepairBucket operation
+type RepairBucketParams struct {
+	DryRun     bool              `json:"dry_run"`
+	RepairJobs []RepairJobParams `json:"repair_jobs"`
+}
+
+// Params suitable for external (eg, HTTP) invocations to describe a specific RepairJob operation
+type RepairJobParams struct {
+	RepairJobType   RepairJobType          `json:"repair_job_type"`
+	RepairJobParams map[string]interface{} `json:"repair_job_params"`
+}
+
+// Given a Couchbase Bucket doc, transform the doc in some way to produce a new doc.
+// Also return a boolean to indicate whether a transformation took place, or any errors occurred.
+type DocTransformer func(doc *document) (transformedDoc *document, transformed bool, err error)
```

Comment:
Should DocTransformer take raw bytes, instead of the marshalled document?  If in future we need to repair something that's breaking document unmarshalling, we won't be able to do it using this pattern.  It does mean that every DocTransformer needs to recode the unmarshal/marshal, but I'm thinking that's not too painful.

Reply:
Done